sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -1225,62 +1225,174 @@ private async Task RunProcessingAsync(CancellationToken cancellationToken)
+        ///
+        /// <remarks>
+        ///   Exceptions encountered when stopping processing for an owned partition will be logged and will result in the error handler
+        ///   being invoked.  They will not be surfaced to callers.  This is intended to be a safe operation consumed
+        ///   as part of the load balancing cycle, which is failure-tolerant.
+        /// </remarks>
+        ///
+        private async Task<bool> TryStopProcessingPartitionAsync(string partitionId,
+                                                                 ProcessingStoppedReason reason,
+                                                                 CancellationToken cancellationToken)
         {
-            // TODO:  Implement this;  current code is just a place holder.
-            if (ActivePartitionProcessors.TryRemove(partitionId, out var processor))
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            Logger.EventProcessorPartitionProcessingStop(partitionId, Identifier, EventHubName, ConsumerGroup);
+
+            var partition = default(TPartition);
+
+            try
             {
-                processor.CancellationSource.Cancel();
+                // If the partition processor is not being tracked or could not be retrieved from the tracking items,
+                // then it cannot be stopped.
+
+                if (!ActivePartitionProcessors.TryRemove(partitionId, out var partitionProcessor))
+                {
+                    return false;
+                }
+
+                // Attempt to stop the processor; any exceptions should be treated as a problem with processing, not
+                // associated with the attempt to stop.
+
+                partition = partitionProcessor.Partition;
 
                 try
                 {
-                    await processor.ProcessingTask.ConfigureAwait(false);
+                    partitionProcessor.CancellationSource.Cancel();
+                    await partitionProcessor.ProcessingTask.ConfigureAwait(false);
                 }
                 catch (TaskCanceledException)
                 {
-                    // nop
+                    // This is expected; no action is needed.
                 }
-                finally
+                catch
                 {
-                    processor.CancellationSource.Dispose();
+                    // The processing task is in a failed state; any logging and dispatching
+                    // to the error handler happened in the processing task before the exception
+                    // was thrown.  All that remains is to override the reason for stopping.
+
+                    reason = ProcessingStoppedReason.OwnershipLost;
                 }
+
+                // Clean up the partition processing resources that need to be disposed.
+
+                partitionProcessor.CancellationSource.Dispose();
```

Comment:
Idle thought - do you think it would be worthwhile to make partitionProcessor Disposable itself, and let it manage disposing the CancellationSource and ProcessingTask?

Reply:
huh... I hadn't thought of that, but that's definitely worth taking a look at.   I'll kick the tires on it tomorrow and see how it looks.

Reply:
It seemed odd to wrap the responsibility for cancellation and awaiting in `DisposeAsync`, so I ended up going with just `Disposable` and letting the processor take responsibility for cleaning up its child resources.